File: 2025/BlogManagement/backend/src/main/java/com/blog/service/BlogPostService.java
package com.blog.service;

import com.blog.model.BlogPost;
import com.blog.repository.BlogPostRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class BlogPostService {

    @Autowired
    private BlogPostRepository blogPostRepository;

    public List<BlogPost> getAllPosts() {
        return blogPostRepository.findAll();
    }

    public BlogPost getPostById(Long id) {
        Optional<BlogPost> post = blogPostRepository.findById(id);
        return post.orElse(null);
    }

    public List<BlogPost> searchPosts(String query) {
        return blogPostRepository.searchPosts(query);
    }

    public List<BlogPost> getPostsByStatus(BlogPost.Status status) {
        return blogPostRepository.findByStatus(status);
    }

    public List<BlogPost> getPostsByCategory(BlogPost.Category category) {
        return blogPostRepository.findByCategory(category);
    }

    public List<BlogPost> getPostsByAuthor(String author) {
        return blogPostRepository.findByAuthor(author);
    }

    public List<BlogPost> getFeaturedPosts() {
        return blogPostRepository.findByFeatured(true);
    }

    public BlogPost createPost(BlogPost post) {
        if (post.getStatus() == null) {
            post.setStatus(BlogPost.Status.DRAFT);
        }
        if (post.getFeatured() == null) {
            post.setFeatured(false);
        }
        return blogPostRepository.save(post);
    }

    public BlogPost updatePost(Long id, BlogPost postDetails) {
        Optional<BlogPost> optionalPost = blogPostRepository.findById(id);
        if (optionalPost.isPresent()) {
            BlogPost post = optionalPost.get();
            post.setTitle(postDetails.getTitle());
            post.setContent(postDetails.getContent());
            post.setAuthor(postDetails.getAuthor());
            post.setCategory(postDetails.getCategory());
            post.setTags(postDetails.getTags());
            post.setStatus(postDetails.getStatus());
            post.setPublicationDate(postDetails.getPublicationDate());
            post.setReadingTime(postDetails.getReadingTime());
            post.setFeatured(postDetails.getFeatured() != null ? postDetails.getFeatured() : false);
            post.setExcerpt(postDetails.getExcerpt());
            return blogPostRepository.save(post);
        }
        return null;
    }

    public boolean deletePost(Long id) {
        if (blogPostRepository.existsById(id)) {
            blogPostRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: 2025/BlogManagement/frontend/src/app/services/blog-post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { BlogPost } from '../models/blog-post';

@Injectable({
  providedIn: 'root'
})
export class BlogPostService {
  private apiUrl = 'http://localhost:8101/api/posts';

  constructor(private http: HttpClient) {}

  getAllPosts(): Observable<BlogPost[]> {
    return this.http.get<BlogPost[]>(this.apiUrl);
  }

  getPostById(id: number): Observable<BlogPost> {
    return this.http.get<BlogPost>(`${this.apiUrl}/${id}`);
  }

  searchPosts(query: string): Observable<BlogPost[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<BlogPost[]>(`${this.apiUrl}/search`, { params });
  }

  getPostsByStatus(status: string): Observable<BlogPost[]> {
    return this.http.get<BlogPost[]>(`${this.apiUrl}/status/${status}`);
  }

  getPostsByCategory(category: string): Observable<BlogPost[]> {
    return this.http.get<BlogPost[]>(`${this.apiUrl}/category/${category}`);
  }

  getPostsByAuthor(author: string): Observable<BlogPost[]> {
    return this.http.get<BlogPost[]>(`${this.apiUrl}/author/${author}`);
  }

  getFeaturedPosts(): Observable<BlogPost[]> {
    return this.http.get<BlogPost[]>(`${this.apiUrl}/featured`);
  }

  createPost(post: BlogPost): Observable<BlogPost> {
    return this.http.post<BlogPost>(this.apiUrl, post);
  }

  updatePost(id: number, post: BlogPost): Observable<BlogPost> {
    return this.http.put<BlogPost>(`${this.apiUrl}/${id}`, post);
  }

  deletePost(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}


